### DISCUSSION TITLE
How to traverse a settingKey

### DISCUSSION THREAD

**Discussion by djx314** [sbt/sbt#7026] [Q&A] [answered]
How to traverse a
```scala
Seq[SettingKey[T]]
```
to a
```scala
SettingKey[Seq[T]]
```

**Comment by eed3si9n** [comment, marked as answer]
sbt internally uses a data structure for arity-generic list called AList for traversing structured (tuples of Initialize[a], Initialize[Task[a]], etc) list-like data. We don't talk much about Seqs since they aren't as useful compared to tuples (HList), but there is a convenience wrapper function called `Def.uniform` that does what you want.

```scala
ThisBuild / scalaVersion     := "2.13.8"
ThisBuild / version          := "0.1.0-SNAPSHOT"

lazy val keyList = settingKey[Seq[String]]("")

lazy val root = (project in file("."))
  .settings(
    name := "uniform demo",
    keyList := Def.uniform(Seq(name, version))(identity).value,
  )
```

https://github.com/sbt/sbt/blob/v1.7.1/internal/util-collection/src/main/scala/sbt/internal/util/Settings.scala#L110-L111

```scala
  def uniform[S, T](inputs: Seq[Initialize[S]])(f: Seq[S] => T): Initialize[T] =
    new Apply[λ[L[x] => List[L[S]]], T](f, inputs.toList, AList.seq[S])
```

**Reply by djx314** [reply]
This suggestion definitely solved my problem.  
Now I try to build a cascade friendly sbt project, one `build.sbt` can only configure it's direct children project's environment and can just get the environment from it's direct children projects.
This makes the configuration of each project depend on the least known information.  
It's the [example](https://github.com/scalax/simple-nat/tree/500dde2492c7c838b651650e0088f35e81314074).

**Comment by SethTisue** [comment]
> there is a convenience wrapper function called Def.uniform that does what you want

the answer is not `traverse`?! 😮😁

**Reply by djx314** [reply]
Yes, I think `Def.uniform` is better. `sbt` use `key.value` to do the things linke `F[_]`. It's something like async await, an it can be deal with mutiply context(`F[_]`, `G[_]`, `A[_]` in one action). And all these is under a `Scope`.  
It's the superset of `F[_]`, injection, and so on.  
So I can understand some unconventional practices.
